src/sentry_sync.c

```diff
@@ -148,31 +182,41 @@ sentry__bgworker_shutdown(sentry_bgworker_t *bgw, uint64_t timeout)
         sentry__mutex_unlock(&bgw->done_signal_lock);
         uint64_t now = sentry__monotonic_time();
         if (now > started && now - started > timeout) {
+            SENTRY_WARN(
+                "background thread failed to shut down cleanly within timeout");
             return 1;
         }
     }
 }
 
-void
+bool
 sentry__bgworker_free(sentry_bgworker_t *bgw)
 {
     if (!bgw) {
-        return;
+        return true;
     }
 
-    assert(!bgw->running && bgw->task_count == 0);
-
-    struct sentry_bgworker_task_s *task = bgw->first_task;
+    sentry__mutex_lock(&bgw->task_lock);
+    sentry_bgworker_task_t *task = bgw->first_task;
     while (task) {
-        struct sentry_bgworker_task_s *next_task = task->next_task;
-        if (task->cleanup_func) {
-            task->cleanup_func(task->data);
-        }
-        sentry_free(task);
+        sentry_bgworker_task_t *next_task = task->next_task;
+        sentry__task_decref(task);
```

Comment:
_nit_: This loop can be moved out of the locked region. Probably irrelevant, since free can be lock-free once we implement refcounting.